fix(cmake): Improve build dependencies and package discovery logic

djgormley opened the pull request:
* **Documentation Build:**
    * Added `add_dependencies(sphinx doxygen)` to ensure Doxygen completes before Sphinx starts. This fixes intermittent build failures related to "Cannot find file: index.xml".

* **Package Discovery:**
    * Set CMake policy `CMP0074` to `NEW` to silence warnings from `find_package(CUDAToolkit)`.
    * Updated HDF5 plugin search paths to correctly locate `libH5Zblosc.so` on Ubuntu 24.04.
    * Removed a premature `find_package(Julia)` call that was causing false "Julia not found" warnings.

* **Console Output:**
    * Corrected message formatting for the LAPACKE discovery and the final build configuration summary.

djgormley committed:
Fix(cmake): Corrected paths and definitions

jbmertens reviewed (approved):
LGTM!

jbmertens commented:
@djgormley let's make sure the linter is passing, then we can merge this in.
(Again, I know the linters don't always produce the best formatting, but still least help keep things consistent.)

djgormley committed:
Applied basic formatting.

djgormley commented:
This also now includes some basic linting on a few misc python files.

jbmertens commented:
> This also now includes some basic linting on a few misc python files.

Thanks - we'll want to make sure the linting agrees with the internal linters (see [tools/lint.sh](https://github.com/kotekan/kotekan/blob/chord/tools/lint.sh)). To run the linters you'll want to have the same versions of `black==19.10b0`, cmake_format (default python 3.12), and clang-format (default ubuntu 24.04 version) installed as the [dockerfile](https://github.com/kotekan/kotekan/blob/chord/tools/docker/24.04/Dockerfile) in this repo.
(You can also start up a container using the dockerfile, and run the linters inside that, if it helps.)

djgormley commented:
> > This also now includes some basic linting on a few misc python files.
> 
> Thanks - we'll want to make sure the linting agrees with the internal linters (see [tools/lint.sh](https://github.com/kotekan/kotekan/blob/chord/tools/lint.sh)). To run the linters you'll want to have the same versions of `black==19.10b0`, cmake_format (default python 3.12), and clang-format (default ubuntu 24.04 version) installed as the [dockerfile](https://github.com/kotekan/kotekan/blob/chord/tools/docker/24.04/Dockerfile) in this repo. (You can also start up a container using the dockerfile, and run the linters inside that, if it helps.)

Thanks. I ran `tools/lint.sh`, and my local setup matches the Dockerfile (Clang 18.1.3, cmake-format 0.6.13) with one critical exception: I am using `black 24.2.0`.

The reason is that the project's `black 19.10b0` is **technically incompatible with Python 3.12**. It relies on deprecated C-extensions (`typed-ast`) and removed libraries (`distutils`) that will fail to build on any clean Python 3.12 environment.

This is a ticking time-bomb for new developer setups and our CI.

I'd strongly recommend we upgrade the Dockerfile to `black 24.2.0` (which is the Ubuntu 24.04 `apt` default) to resolve this incompatibility. I also agree we should add version checks to `lint.sh` to catch these mismatches early.

(This is actually why I like to run the tools locally outside of Docker—it's a great way to surface hidden build and dependency issues like this.)

jbmertens commented:
> Thanks. I ran `tools/lint.sh`, and my local setup matches the Dockerfile (Clang 18.1.3, cmake-format 0.6.13) with one critical exception: I am using `black 24.2.0`.
> 
> The reason is that the project's `black 19.10b0` is **technically incompatible with Python 3.12**. It relies on deprecated C-extensions (`typed-ast`) and removed libraries (`distutils`) that will fail to build on any clean Python 3.12 environment.
> 
> This is a ticking time-bomb for new developer setups and our CI.
> 
> I'd strongly recommend we upgrade the Dockerfile to `black 24.2.0` (which is the Ubuntu 24.04 `apt` default) to resolve this incompatibility. I also agree we should add version checks to `lint.sh` to catch these mismatches early.
> 
> (This is actually why I like to run the tools locally outside of Docker—it's a great way to surface hidden build and dependency issues like this.)

Sure - makes sense. If these are the only python changes, that is surprisingly minimal!
Indeed, can you update the black version in the Dockerfile as well? That should cause the CI to lint "correctly".

djgormley commented:
> > Thanks. I ran `tools/lint.sh`, and my local setup matches the Dockerfile (Clang 18.1.3, cmake-format 0.6.13) with one critical exception: I am using `black 24.2.0`.
> > The reason is that the project's `black 19.10b0` is **technically incompatible with Python 3.12**. It relies on deprecated C-extensions (`typed-ast`) and removed libraries (`distutils`) that will fail to build on any clean Python 3.12 environment.
> > This is a ticking time-bomb for new developer setups and our CI.
> > I'd strongly recommend we upgrade the Dockerfile to `black 24.2.0` (which is the Ubuntu 24.04 `apt` default) to resolve this incompatibility. I also agree we should add version checks to `lint.sh` to catch these mismatches early.
> > (This is actually why I like to run the tools locally outside of Docker—it's a great way to surface hidden build and dependency issues like this.)
> 
> Sure - makes sense. If these are the only python changes, that is surprisingly minimal! Indeed, can you update the black version in the Dockerfile as well? That should cause the CI to lint "correctly".

This has inspired me to properly update the docker.  I'll probably close this PR and start a new one.

> > Thanks. I ran `tools/lint.sh`, and my local setup matches the Dockerfile (Clang 18.1.3, cmake-format 0.6.13) with one critical exception: I am using `black 24.2.0`.
> > The reason is that the project's `black 19.10b0` is **technically incompatible with Python 3.12**. It relies on deprecated C-extensions (`typed-ast`) and removed libraries (`distutils`) that will fail to build on any clean Python 3.12 environment.
> > This is a ticking time-bomb for new developer setups and our CI.
> > I'd strongly recommend we upgrade the Dockerfile to `black 24.2.0` (which is the Ubuntu 24.04 `apt` default) to resolve this incompatibility. I also agree we should add version checks to `lint.sh` to catch these mismatches early.
> > (This is actually why I like to run the tools locally outside of Docker—it's a great way to surface hidden build and dependency issues like this.)
> 
> Sure - makes sense. If these are the only python changes, that is surprisingly minimal! Indeed, can you update the black version in the Dockerfile as well? That should cause the CI to lint "correctly".

Going to close this PR down and make a new one centered on and updated build.
